### PR TITLE
Soumyo/ec2 eip resource

### PR DESCRIPTION
**Pull request by soumyo13**
### Description

This pull request contains the resource for EC2 EIP/Elastic IP.

### Issues Resolved
This is a completely new resource. So there are no issues to be resolved.

### Check List
Please fill box or appropriate ([x]) or mark N/A.
- [x] New functionality includes integration tests/controls
- [x] New Terraform resources
- [x] Documentation provided or updated for resources 
- [x] All Integration Tests pass
- [x] All Unit Tests pass
- [x] `rake lint` passes
- [x] All commits have been signed-off for the Developer Certificate of Origin. See <https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco>

### CODE REVIEW

File: libraries/aws_ec2_eips.rb [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+  def initialize(opts = {})
+    super(opts)
+    validate_parameters
+    @table = fetch_data
+  end
+
+  def fetch_data
+    catch_aws_errors do
+      @addrs = @aws.compute_client.describe_addresses
```
Comment (sathish-progress) [inline libraries/aws_ec2_eips.rb:36]: does it support pagination?

File: libraries/aws_backend.rb [hunk 2]
```diff
@@ -172,6 +172,26 @@ def sts_client
...
+    aws_client(Aws::Athena::Client)
+  end
+
+  def autoscaling_client
+    aws_client(Aws::AutoScaling::Client)
+  end
+
+  def batch_client
+    aws_client(Aws::Batch::Client)
+  end
```
Comment (sathish-progress) [inline libraries/aws_backend.rb:194]: why do we need this?

Comment (soumyo13) [reply libraries/aws_backend.rb:194]: These are required when we will be writing the other resources?

Comment (sathish-progress) [reply libraries/aws_backend.rb:194]: Okay, please remove them if it is not relevant to this PR..

Comment (soumyo13) [reply libraries/aws_backend.rb:194]: OK

File: libraries/aws_ec2_eips.rb (hunk 1 above)
Comment (soumyo13) [reply libraries/aws_ec2_eips.rb:36]: Not Required... https://docs.aws.amazon.com/sdk-for-ruby/v3/api/Aws/EC2/Client.html#describe_addresses-instance_method

File: docs/resources/aws_ec2_eip.md [hunk 3]
```diff
@@ -0,0 +1,94 @@
+---
+title: About the aws_ec2_eip Resource
+platform: aws
+---
+
+# aws\_eip
+
+Use the `aws_eip` InSpec audit resource to test properties of a single specific Elastic IP (EIP). 
+
+A Elastic IP (EIP) is uniquely identified by the Public IP (e.g. 18.192.107.8)
```
Comment (IanMadd) [inline docs/resources/aws_ec2_eip.md:10]: There are several IP addresses reserved for documentation.
```suggestion
An Elastic IP (EIP) is uniquely identified by the public IPv4 address, for example `192.0.2.0`.
```

File: docs/resources/aws_ec2_eip.md [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+## Syntax
+
+Ensure that a Public IP exists.
+
+    # Find a Public IP by public_ip
+    describe aws_ec2_eip(public_ip: '18.192.107.8') do
```
Comment (IanMadd) [inline docs/resources/aws_ec2_eip.md:17]:
```suggestion
    describe aws_ec2_eip(public_ip: '192.0.2.0') do
```

File: docs/resources/aws_ec2_eip.md [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+      it { should exist }
+    end
+    
+    # Hash syntax for public_ip
+    describe aws_ec2_eip(public_ip: '18.192.107.8') do
```
Comment (IanMadd) [inline docs/resources/aws_ec2_eip.md:22]:
```suggestion
    describe aws_ec2_eip(public_ip: '192.0.2.0') do
```

File: docs/resources/aws_ec2_eip.md [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+      it { should exist }
+    end
+
+## Parameters
+
+This resource expects the Public IP as a parameter.
```
Comment (IanMadd) [inline docs/resources/aws_ec2_eip.md:28]:
```suggestion
`public_ip` _(required)_

This resource requires the public IPv4 address as a string.
```

File: docs/resources/aws_ec2_eip.md [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+
+See also the [AWS documentation on Elastic IP (EIP)](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-eip.html).
```
Comment (IanMadd) [inline docs/resources/aws_ec2_eip.md:30]:
```suggestion
For additional information, see the [AWS documentation on Elastic IP (EIP)](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-eip.html).
```

File: docs/resources/aws_ec2_eip.md [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+| domain | Indicates whether the address is for use in EC2-Classic (standard) or in a VPC (vpc). |
+| network_interface_id | The ID of the network interface that the address is associated with, if any. |
+| network_interface_owner_id | The AWS account ID of the owner. |
+| private_ip_address | The private IP address associated with the Elastic IP address. |
+| public_ipv_4_pool | The Elastic IPV4 pool address. |
+| network_border_group | A unique set of Availability Zones, Local Zones, or Wavelength Zones from where AWS advertises IP addresses. |
+
+## Examples
+
+##### Ensure a Public IP is available.
```
Comment (IanMadd) [inline docs/resources/aws_ec2_eip.md:49]:
```suggestion
### Ensure a Public IP is available.

```

File: docs/resources/aws_ec2_eip.md [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+    describe aws_ec2_eip(public_ip: '18.192.107.8') do
+      its('public_ip') { should eq '18.192.107.8' }
```
Comment (IanMadd) [inline docs/resources/aws_ec2_eip.md:51]:
```suggestion
    describe aws_ec2_eip(public_ip: '192.0.2.0') do
      its('public_ip') { should eq '192.0.2.0' }
```

File: docs/resources/aws_ec2_eip.md [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+    end
+
+##### Ensure that the domain is vpc or standard.
```
Comment (IanMadd) [inline docs/resources/aws_ec2_eip.md:54]:
```suggestion
### Ensure that the domain is vpc or standard.

```

File: docs/resources/aws_ec2_eip.md [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+    describe aws_ec2_eip(public_ip: '18.192.107.8') do
```
Comment (IanMadd) [inline docs/resources/aws_ec2_eip.md:55]:
```suggestion
    describe aws_ec2_eip(public_ip: '192.0.2.0') do
```

File: docs/resources/aws_ec2_eip.md [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+        its('domain') { should eq 'vpc' }
+    end
+
+## Matchers
+
+This InSpec audit resource has the following special matchers. For a full list of available matchers, please visit our [Universal Matchers page](https://www.inspec.io/docs/reference/matchers/).
+
+The controls will pass if the describe returns at least one result.
```
Comment (IanMadd) [inline docs/resources/aws_ec2_eip.md:63]:
```suggestion
The controls will pass if the `describe` method returns at least one result.
```

File: docs/resources/aws_ec2_eip.md [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+
+#### exist
```
Comment (IanMadd) [inline docs/resources/aws_ec2_eip.md:65]:
```suggestion
### exist
```

File: docs/resources/aws_ec2_eip.md [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+
+Use `should_not` to test the entity should not exist.
```
Comment (IanMadd) [inline docs/resources/aws_ec2_eip.md:67]:
```suggestion
Use `should` to test that the entity exists.
```

File: docs/resources/aws_ec2_eip.md [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+
+    describe aws_ec2_eip(public_ip: '18.192.107.8') do
```
Comment (IanMadd) [inline docs/resources/aws_ec2_eip.md:69]:
```suggestion
    describe aws_ec2_eip(public_ip: '192.0.2.0') do
```

File: docs/resources/aws_ec2_eip.md [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+      it { should exist }
+    end
+
```
Comment (IanMadd) [inline docs/resources/aws_ec2_eip.md:72]:
```suggestion

Use `should_not` to test that the entity does not exist.

```

File: docs/resources/aws_ec2_eip.md [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+#### exist
+
+Use `should_not` to test the entity should not exist.
+
+    describe aws_ec2_eip(public_ip: '18.192.107.8') do
+      it { should exist }
+    end
+      
+    describe aws_vpc_endpoint(public_ip: '18.192.101.8') do
```
Comment (IanMadd) [inline docs/resources/aws_ec2_eip.md:73]:
```suggestion
    describe aws_vpc_endpoint(public_ip: '192.0.2.0') do
```

File: docs/resources/aws_ec2_eip.md [hunk 18]
```diff
@@ -0,0 +1,94 @@
...
+      it { should_not exist }
+    end
+
+#### be_available
```
Comment (IanMadd) [inline docs/resources/aws_ec2_eip.md:77]:
```suggestion
### be_available
```

File: docs/resources/aws_ec2_eip.md [hunk 19]
```diff
@@ -0,0 +1,94 @@
...
+
+Checks if the Addresses is in available state.
+Use `should_not` to test the entity should not exist.
```
Comment (IanMadd) [inline docs/resources/aws_ec2_eip.md:80]:
```suggestion
Check if the IP address is available.
```

File: docs/resources/aws_ec2_eip.md [hunk 20]
```diff
@@ -0,0 +1,94 @@
...
+
+    describe aws_ec2_eip(public_ip: '18.192.107.8') do
```
Comment (IanMadd) [inline docs/resources/aws_ec2_eip.md:82]:
```suggestion
    describe aws_ec2_eip(public_ip: '192.0.2.0') do
```

File: docs/resources/aws_ec2_eip.md [hunk 21]
```diff
@@ -0,0 +1,94 @@
...
+      it { should be_available }
+    end
+
```
Comment (IanMadd) [inline docs/resources/aws_ec2_eip.md:85]:
```suggestion

Use `should_not` to test an IP address that should not exist.

```

File: docs/resources/aws_ec2_eip.md [hunk 22]
```diff
@@ -0,0 +1,94 @@
...
+      it { should be_available }
+    end
+      
+    describe aws_ec2_eip(public_ip: '18.192.101.8') do
+      it { should_not be_available }
+    end
+
+## AWS Permissions
+
+Your [Principal](https://docs.aws.amazon.com/IAM/latest/UserGuide/intro-structure.html#intro-structure-principal) will need the `ec2:DescribeAddresses` action with Effect set to Allow.
```
Comment (IanMadd) [inline docs/resources/aws_ec2_eip.md:92]:
```suggestion
Your [Principal](https://docs.aws.amazon.com/IAM/latest/UserGuide/intro-structure.html#intro-structure-principal) will need the `ec2:DescribeAddresses` action with `Effect` set to `Allow`.
```

File: docs/resources/aws_ec2_eip.md [hunk 23]
```diff
@@ -0,0 +1,94 @@
...
+
+You can find detailed documentation at [Actions, Resources, and Condition Keys for Amazon EC2](https://docs.aws.amazon.com/IAM/latest/UserGuide/list_amazonec2.html).
```
Comment (IanMadd) [inline docs/resources/aws_ec2_eip.md:94]:
```suggestion
See the [Actions, Resources, and Condition Keys for Amazon EC2](https://docs.aws.amazon.com/IAM/latest/UserGuide/list_amazonec2.html) documentation for additional information.
```

File: docs/resources/aws_ec2_eips.md [hunk 24]
```diff
@@ -0,0 +1,81 @@
+---
+title: About the aws_ec2_eips Resource
+platform: aws
+---
+
+# aws\_eips
+
+Use the `aws_eips` InSpec audit resource to test properties of some or all AWS Elastic IPs.
```
Comment (IanMadd) [inline docs/resources/aws_ec2_eips.md:8]:
```suggestion
Use the `aws_eips` InSpec audit resource to test properties of some or all AWS Elastic IP addresses.
```

File: docs/resources/aws_ec2_eips.md [hunk 25]
```diff
@@ -0,0 +1,81 @@
...
+
+## Syntax
+
+Ensure that a Public IP exists.
```
Comment (IanMadd) [inline docs/resources/aws_ec2_eips.md:12]:
```suggestion
Verify that a public IP address exists.
```

File: docs/resources/aws_ec2_eips.md [hunk 26]
```diff
@@ -0,0 +1,81 @@
...
+
+    # Find a Public IP by public_ip
```
Comment (IanMadd) [inline docs/resources/aws_ec2_eips.md:14]:
```suggestion
```

File: docs/resources/aws_ec2_eips.md [hunk 27]
```diff
@@ -0,0 +1,81 @@
...
+    describe aws_ec2_eip(public_ip: '18.192.107.8') do
```
Comment (IanMadd) [inline docs/resources/aws_ec2_eips.md:15]:
```suggestion
    describe aws_ec2_eip(public_ip: '192.0.2.0') do
```

File: docs/resources/aws_ec2_eips.md [hunk 28]
```diff
@@ -0,0 +1,81 @@
...
+      it { should exist }
+    end
+
+An `aws_eips` resource block uses an optional filter to select a group of Elastic IPs and then tests that group.
```
Comment (IanMadd) [inline docs/resources/aws_ec2_eips.md:19]:
```suggestion
An `aws_eips` resource block uses an optional filter to select a group of Elastic IPs and then test that group.
```

File: docs/resources/aws_ec2_eips.md [hunk 29]
```diff
@@ -0,0 +1,81 @@
...
+
+## Parameters
+
+This resource does not expect any parameters.
+
+See also the [AWS documentation on Elastic IP (EIP)](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-eip.html).
```
Comment (IanMadd) [inline docs/resources/aws_ec2_eips.md:25]:
```suggestion
See the [AWS documentation on Elastic IP (EIP)](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-eip.html) for additional information.
```

File: libraries/aws_ec2_eip.rb [hunk 30]
```diff
@@ -0,0 +1,49 @@
...
+
+  example "
+    describe aws_ec2_eip(public_ip: '10.1.1.10') do
+      it { should eq '10.1.1.10' }
+    end
+
+    describe aws_ec2_eip(public_ip: '10.1.1.10') do
+      it { should exits }
+    end
+  "
```
Comment (sathish-progress) [inline libraries/aws_ec2_eip.rb:17]: add one more example for this, 

```
describe aws_ec2_eip('10.1.1.10') do
      it { should exits }
    end
```

File: libraries/aws_ec2_eip.rb [hunk 31]
```diff
@@ -0,0 +1,49 @@
...
+    catch_aws_errors do
+      resp = @aws.compute_client.describe_addresses({ public_ips: [opts[:public_ip]] })
+      @addresses = resp.addresses[0].to_h
+      create_resource_methods(@addresses)
+    end
+  end
+
+  def id
+    return nil unless exists?
+    @volume[:public_ip]
```
Comment (sathish-progress) [inline libraries/aws_ec2_eip.rb:35]: copy-paste error!

File: test/integration/verify/controls/aws_ec2_eip.rb [hunk 32]
```diff
@@ -0,0 +1,37 @@
...
+aws_allocation_id = attribute("aws_allocation_id", default: "", description: "The allocation ID for the address.")
+aws_association_id = attribute("aws_association_id", default: "", description: "The association ID for the address.")
+aws_domain = attribute("aws_domain", default: "vpc", description: "Indicates whether the address is for use in EC2-Classic (standard) or in a VPC (vpc).")
+aws_network_interface_id = attribute("aws_network_interface_id", default: "", description: "The ID of the network interface that the address is associated with, if any.")
+aws_network_interface_owner_id = attribute("aws_network_interface_owner_id", default: "", description: "The AWS account ID of the owner.")
+aws_private_ip_address = attribute("aws_private_ip_address", default: "", description: "The private IP address associated with the Elastic IP address.")
+aws_public_ipv_4_pool = attribute("aws_public_ipv_4_pool", default: "", description: "The Elastic IPV4 pool address.")
+aws_network_border_group = attribute("aws_network_border_group", default: "", description: "A unique set of Availability Zones, Local Zones, or Wavelength Zones from where AWS advertises IP addresses.")
```
Comment (sathish-progress) [inline test/integration/verify/controls/aws_ec2_eip.rb:10]: None of the values except `aws_instance_id` is present in `outputs.tf`.. please assign and test once!

File: libraries/aws_ec2_eip.rb (hunk 30 above)
Comment (sathish-progress) [reply libraries/aws_ec2_eip.rb:17]: ??

Comment (soumyo13) [reply libraries/aws_ec2_eip.rb:17]: Already examples are there in the documentation.

Comment (soumyo13) [reply libraries/aws_ec2_eip.rb:17]: @sathish-progress  if someone wants to see the examples then one should go to the documentation and see how it works. Right?

Comment (soumyo13) [reply libraries/aws_ec2_eip.rb:17]: Done

File: libraries/aws_auto_scaling_groups.rb [hunk 33]
```diff
@@ -45,11 +45,15 @@ def fetch_data
                          health_check_type:         group[:health_check_type] }]
 
         group[:tags].map { |tag|
-          config_rows.push({ 'resource_id':          tag.resource_id,
```
Comment (sathish-progress) [inline libraries/aws_auto_scaling_groups.rb:48]: please remove this files from here..

**Comment by sathish-progress** [comment]
also add resource doc links to readme

File: test/integration/build/outputs.tf [hunk 34]
```diff
@@ -294,6 +294,47 @@ output "aws_guardduty_detector_publishing_frequency" {
   value = aws_guardduty_detector.detector_1.0.finding_publishing_frequency
 }
 
+output "aws_public_ip" {
+  value = aws_eip.aws_eip_1.public_ip
+}
+
+//noinspection HILUnresolvedReference
```
Comment (sathish-progress) [inline test/integration/build/outputs.tf:301]: all with this comment is unresolved!!

**Review by ppradhan9 — APPROVED** [review]
Thanks! @soumyo13

**Pull request merged**